grafana/grafana
# Graphite: Interpolate Graphite alerts with tags

**What this PR does / why we need it**:

This is to refactor Graphite backend queries to data frames. It also adds support for parsing tags which allows to use them as variables in templates (#32977). It will conflict with data frame conversion applied in #35798 but the implementation differs a bit so we might need to do some mix & match :)

In this branch:
* Added handling for null/nil values. Some points might be missing when coming from Graphite. In previous implementation with DataPoints they were mapped to 0 instead of `nil`.
* Invalid timestamp results with an error.
* Also added some unit tests
* Added support for tags (returned by Graphite in a separate structure)

**Which issue(s) this PR fixes**:



Fixes #32977
Subtask of #22918 
Conflicts with #35798

**Special notes for your reviewer**:

How to test alerting?

Run dev environment:

```
make devenv sources=graphite
make run
```

Pass random data to Graphite running [ ./graphite-alerting.sh](https://gist.github.com/ifrost/29e5b4041b0362e968b82ff459360065#file-graphite-alerting-sh)

Configure SMTP.  For example you can use  [mailslurper ](https://github.com/mailslurper/mailslurper) with following custom.ini config:

```
[smtp]
enabled = true
host = localhost:2500
user =
password =
cert_file =
key_file =
skip_verify = false
from_address = admin@grafana.localhost
from_name = Grafana
ehlo_identity =
startTLS_policy =
```

Create Alert notification channel:

<img width="300" alt="Screenshot 2021-06-17 at 15 11 28" src="https://user-images.githubusercontent.com/745532/122406262-ff3ec780-cf80-11eb-9471-c737535645ba.png">

Create a dashboard with the following query (make sure to use `tag:environment` when selecting the tag name, or use raw query: `(seriesByTag('environment=prod')`):

<img width="300" alt="Screenshot 2021-06-17 at 15 00 46" src="https://user-images.githubusercontent.com/745532/122405661-850e4300-cf80-11eb-85af-8bab454f7c1b.png">

Create an alert with variables:

<img width="500" alt="Screenshot 2021-06-17 at 15 05 52" src="https://user-images.githubusercontent.com/745532/122406389-1a113c00-cf81-11eb-80c9-156114a2113e.png">

Go back to the script that is sending random data and increase the baseline above 60 (using up/down arrows). Metric values should go up triggering an alert:

<img width="744" alt="Screenshot 2021-06-17 at 15 54 08" src="https://user-images.githubusercontent.com/745532/122410576-54300d00-cf84-11eb-958d-0fe2fc82683c.png">

In mailslurper (http://localhost:8080) you should see:
<img width="300" alt="Screenshot 2021-06-17 at 15 07 18" src="https://user-images.githubusercontent.com/745532/122406352-12519780-cf81-11eb-9d75-0d1e66a23c52.png">

⚠️ When alert goes back to "ok" stage the interpolation won't work due to #30606
